[18.0][MIG] stock_move_backdating

bosd opened the pull request.

tafaRU committed:
[ADD] stock_move_backdating

tafaRU committed:
[FIX]stock_move_backdating: stock_view.xml

eLBati committed:
[FIX] authors

tafaRU committed:
[REF] stock_partial_picking_line

tafaRU committed:
[FIX] changed view

eLBati committed:
[imp] use back date for accounting

eLBati committed:
[imp] orm.Model

eLBati committed:
[imp] use move date and related period

eLBati committed:
[add] missing account.py

tafaRU committed:
[FIX] stock_partial_picking_line

tafaRU committed:
[ADD] override to _partial_move_for and to do_partial in stock.partial.picking

eLBati committed:
[fix] wrong create return value

tafaRU committed:
[FIX] remove pdb

tafaRU committed:
[ADD] AUTHORS and copyright

tafaRU committed:
[ADD] mail address to AUTHORS.txt

tafaRU committed:
[FIX] PEP 8 and OpenERP syntax compliance

eLBati committed:
[FIX] using right date: move.date_backdating

tafaRU committed:
[FIX]replaced m by move and avoid extra spaces

eLBati committed:
[FIX] when move.date_backdating is not set

eLBati committed:
[IMP] stock_move_backdating and stock_picking_invoice_link descriptions

gurneyalex committed:
mark unported module as uninstallable

gurneyalex committed:
Update travis configuration

use sudo: false to enable container build
use cache: pip to cache pip packages

archetipo committed:
porting of the module [FIX] tests [FIX] translation

archetipo committed:
update quants date [FIX] test

eLBati committed:
FIX when using module with user != admin. Test added

sbidoul committed:
[MOV] move addons out of __unported__ (they remain not installable)

archetipo committed:
README [FIX] Travis.yaml [FIX] rebased with 8.0

archetipo committed:
author

archetipo committed:
Wizard date comparison [RM] useless inheritance class stock_quant and method _create_account_move_line [FIX] upgrade quant_ids date in a single call

archetipo committed:
stock logistic workflow available addons

archetipo committed:
travis remarks

archetipo committed:
version number

archetipo committed:
Copyright in py module

oca-transbot committed:
OCA Transbot updated translations from Transifex

pedrobaeza committed:
[MIG] Rename manifest files

tafaRU committed:
stock_move_backdating module (#252)

* Apply some cosmetic changes according to OCA guidelines

* Add date_backdating to stock.transfer.details and set date_done of stock.picking as the youngest date among moves

* Update test accordingly to last changes

* Update i18n/it.po accordingly to last changes

* Apply some cosmetic changes: copyright and OCA url

* [IMP] stock_move_backdating: decription

* FIX stock_move_backdating: avoid to use future date anyway, if forced in web interface

tafaRU committed:
stock_move_backdating: fix TestStockMoveBackdating

oca-transbot committed:
OCA Transbot updated translations from Transifex

sbidoul committed:
fix website url in manifest

tafaRU committed:
[MIG] stock_move_backdating: Migration to 10.0

tafaRU committed:
[ADD] stock_move_backdating: add 'Fill Actual Movement Date' button

tafaRU committed:
[FIX] stock_move_backdating: avoid searching in a loop and filter only done moves

tafaRU committed:
[FIX] stock_move_backdating: prevent check_date for empty date

tafaRU committed:
[IMP] stock_move_backdating: add check_date to FillDateBackdating wizard

pedrocs-exo committed:
Translated using Weblate (Portuguese)

Currently translated at 88.2% (15 of 17 strings)

Translation: stock-logistics-workflow-10.0/stock-logistics-workflow-10.0-stock_move_backdating
Translate-URL: https://translation.odoo-community.org/projects/stock-logistics-workflow-10-0/stock-logistics-workflow-10-0-stock_move_backdating/pt/

SirTakobi committed:
[MIG] stock_move_backdating: Migration to 12.0

SirTakobi committed:
[IMP] stock_move_backdating: Update source quant date

SirTakobi committed:
[IMP] stock_move_backdating: Show date in picking

SirTakobi committed:
[I18N] stock_move_backdating: Update Italian and POT

SirTakobi committed:
[IMP] stock_move_backdating: Allow to backdate inventory

SirTakobi committed:
[IMP] stock_move_backdating: Performance

Co-authored-by: Denis Roussel <denis.roussel@acsone.eu>

Commit:
[UPD] Update stock_move_backdating.pot

OCA-git-bot committed:
[UPD] README.rst

OCA-git-bot committed:
stock_move_backdating 12.0.1.0.1

mymage committed:
Translated using Weblate (Italian)

Currently translated at 95.4% (21 of 22 strings)

Translation: stock-logistics-workflow-12.0/stock-logistics-workflow-12.0-stock_move_backdating
Translate-URL: https://translation.odoo-community.org/projects/stock-logistics-workflow-12-0/stock-logistics-workflow-12-0-stock_move_backdating/it/

ps-tubtim committed:
[IMP] stock_move_backdating: black, isort, prettier

ps-tubtim committed:
[MIG] stock_move_backdating: Migration to 15.0

OCA-git-bot committed:
[UPD] README.rst

mymage committed:
Translated using Weblate (Italian)

Currently translated at 100.0% (22 of 22 strings)

Translation: stock-logistics-workflow-12.0/stock-logistics-workflow-12.0-stock_move_backdating
Translate-URL: https://translation.odoo-community.org/projects/stock-logistics-workflow-12-0/stock-logistics-workflow-12-0-stock_move_backdating/it/

toita86 committed:
[IMP] stock_move_backdating: black, isort, prettier

toita86 committed:
[MIG] stock_move_backdating: Migration to 14.0

Commit:
[UPD] Update stock_move_backdating.pot

OCA-git-bot committed:
[BOT] post-merge updates

Saran440 committed:
[MIG] stock_move_backdating: Migration to 16.0

Saran440 committed:
[IMP] stock_move_backdating: add scrap back date & update readme

bosd committed:
[IMP] stock_move_backdating: pre-commit auto fixes

bosd committed:
[MIG] stock_move_backdating: Migration to 18.0

bosd committed:
[MIG] stock_move_backdating: Migration to 18.0

map-240221 committed:
[17.0][MIG] stock_move_backdating: Migration to 17.0

map-240221 committed:
[DEL] stock_move_backdating: remove unused file

map-240221 committed:
[IMP] stock_move_backdating: variable adjustments

Saran440 committed:
[FIX] stock_move_backdating: update date backdating when stock is backdate only

map-240221 committed:
[UPD] stock_move_backdating: increase test coverage

bosd commented:
@map-240221 , @Saran440 Can you please review?

Saran440 commented:
@bosd Can you check red test? may be it error from function check_date()

bosd commented:
> Can you check red test?

CI is green on the unit tests without the rebel modules. Also passing locally as wel..

Dranyel-Bosd reviewed (approved):
LGTM :100:

Saran440 reviewed (changes requested):
@bosd We should migrate `stock_move_backdating` module only.
I see flie change has `.copier-answers.yml, test.yml, .pre-commit.config.yaml`

Can you check please?

bosd commented:
Yes, that is intentional to rule out the rebel modules in the unit tests.

Earlier I reverted those changes and it made the test fail.

This module is just not " friends" with everyone 🙄

Saran440 commented:
Functional Test :+1: 

@bosd Could you try reverting the code to check for the error? I’ll take a look.

bosd commented:
> @bosd Could you try reverting the code to check for the error? I’ll take a look.
@Saran440 I'm not sure what you mean by this.

The copier update of the dotfiles just made 2 separate test stuies.
So now it is possible to acces the isolated tests:

![image](https://github.com/user-attachments/assets/e08cc54e-da26-4a77-92f0-1a6cf09ab02d)
[OCB](https://github.com/OCA/stock-logistics-workflow/actions/runs/15684787215/job/44185108795?pr=1921)
[odoo](https://github.com/OCA/stock-logistics-workflow/actions/runs/15684787215/job/44185108800?pr=1921)

as well as the one with rebel modules:
![image](https://github.com/user-attachments/assets/ca4c191d-d601-4156-b0d7-033cd78c262d)
[OCB](https://github.com/OCA/stock-logistics-workflow/actions/runs/15684787215/job/44185108801?pr=1921)
[ODOO](https://github.com/OCA/stock-logistics-workflow/actions/runs/15684787215/job/44185108811?pr=1921)

So if you want to work on improving the compatability with rebel modules. The last two links are a good start.

I have not looked into this a lot. 
But it could be that your previous fix
https://github.com/OCA/stock-logistics-workflow/pull/1921/commits/0fae21a05006d208cad9f5df1f7090464e605bc9
Is no longer applicable to V18 (potentiall V17+) because of the changes in odoo core, how they handle openen ended domain.

(Sorry could not quickly find a link to the odoo core where they made these changes).
Edit: Here it is: https://github.com/odoo/odoo/commit/450f5c9

micheledic commented on stock_move_backdating/models/stock_move.py:
```diff
@@ -0,0 +1,110 @@
+                    order.company_id,
+                    date_backdating,
+                    round=False,
+                )
```
```
)
price_unit = {self.env['stock.lot']: price_unit}
```

micheledic commented on stock_move_backdating/models/stock_move.py:
```diff
@@ -0,0 +1,110 @@
+                    order.company_id,
+                    date_backdating,
+                    round=False,
+                )
```
```
)
price_unit = {self.env['stock.lot']: price_unit}
```

at line 40

bobslee replied:
[stock_move_backdating.patch](https://github.com/user-attachments/files/26210495/stock_move_backdating.patch)

This is a patch for `_get_price_unit`, to align with `stock_account/models/stock_move.py`.
The tests still succeed.

@bosd Please apply the patch. Thank you.

micheledic reviewed (changes requested):
_get_price_unit should return a dict with key = stock.lot and the value the price unit

bosd committed:
IMP Suggestion from micheledic

andreampiovesana commented:
@micheledic is ok?

TheerayutEncoder commented:
After installed the module, I test with create new Inventory Adjustment
<img width="1289" height="238" alt="image" src="https://github.com/user-attachments/assets/202070f6-1bb0-4320-95d9-ee6e0ef7712d" />

Then add `Counted Quantities` and `Actual Inventory Date`

When I click save and click apply
<img width="1278" height="270" alt="image" src="https://github.com/user-attachments/assets/0f0d0420-49a1-48fe-a922-d2ea0e0dbf39" />

Error as below
```python
RPC_ERROR

Odoo Server Error

Occured on localhost:18069 on model stock.quant on 2026-01-06 06:36:37 GMT

Traceback (most recent call last):
  File "/opt/odoo/custom/src/odoo/odoo/http.py", line 2166, in _transactioning
    return service_model.retrying(func, env=self.env)
  File "/opt/odoo/custom/src/odoo/odoo/service/model.py", line 156, in retrying
    result = func()
  File "/opt/odoo/custom/src/odoo/odoo/http.py", line 2133, in _serve_ir_http
    response = self.dispatcher.dispatch(rule.endpoint, args)
  File "/opt/odoo/custom/src/odoo/odoo/http.py", line 2381, in dispatch
    result = self.request.registry['ir.http']._dispatch(endpoint)
  File "/opt/odoo/custom/src/odoo/odoo/addons/base/models/ir_http.py", line 333, in _dispatch
    result = endpoint(**request.params)
  File "/opt/odoo/custom/src/odoo/odoo/http.py", line 754, in route_wrapper
    result = endpoint(self, *args, **params_ok)
  File "/opt/odoo/auto/addons/web/controllers/dataset.py", line 42, in call_button
    action = call_kw(request.env[model], method, args, kwargs)
  File "/opt/odoo/custom/src/odoo/odoo/api.py", line 535, in call_kw
    result = getattr(recs, name)(*args, **kwargs)
  File "/opt/odoo/auto/addons/stock/models/stock_quant.py", line 474, in action_apply_inventory
    self._apply_inventory()
  File "/opt/odoo/auto/addons/stock_move_backdating/models/stock_quant.py", line 57, in _apply_inventory
    super(StockQuant, inventory_ctx)._apply_inventory()
  File "/opt/odoo/auto/addons/stock_account/models/stock_quant.py", line 74, in _apply_inventory
    super(StockQuant, inventories)._apply_inventory()
  File "/opt/odoo/auto/addons/stock/models/stock_quant.py", line 1032, in _apply_inventory
    moves._action_done()
  File "/opt/odoo/auto/addons/stock_move_backdating/models/stock_move.py", line 98, in _action_done
    moves_todo = self._backdating_action_done(
  File "/opt/odoo/auto/addons/stock_move_backdating/models/stock_move.py", line 74, in _backdating_action_done
    move_todo = super()._action_done(cancel_backorder=cancel_backorder)
  File "/opt/odoo/auto/addons/stock_account/models/stock_move.py", line 369, in _action_done
    stock_valuation_layers |= getattr(todo_valued_moves, '_create_%s_svl' % valued_type)()
  File "/opt/odoo/auto/addons/stock_account/models/stock_move.py", line 220, in _create_in_svl
    svl_vals_list = self._get_in_svl_vals(forced_quantity)
  File "/opt/odoo/auto/addons/stock_account/models/stock_move.py", line 535, in _get_in_svl_vals
    vals = [move.product_id._prepare_in_svl_vals(sum(quantities.values()), abs(unit_cost[self.env['stock.lot']]))]
TypeError: bad operand type for abs(): 'dict'

The above server error caused the following client error:
RPC_ERROR: Odoo Server Error
    RPC_ERROR
        at makeErrorFromResponse (http://localhost:18069/web/assets/ca9ac20/web.assets_web.min.js:3165:165)
        at XMLHttpRequest.<anonymous> (http://localhost:18069/web/assets/ca9ac20/web.assets_web.min.js:3170:13)
```

bobslee commented:
Regarding the previous comment: https://github.com/OCA/stock-logistics-workflow/pull/1921#issuecomment-3713327327
I don't experience that error.

Commit:
[FIX] stock_move_backdating: fix _update_available_quantity signature

The override renamed the 'quantity' parameter to 'reserved_quantity',
causing all stock receipts to set reserved_quantity instead of on-hand
quantity on quants. This resulted in on-hand=0 for every validated
receipt while reserved_quantity was incorrectly set.

The base signature is:
  _update_available_quantity(product_id, location_id, quantity=False,
    reserved_quantity=False, ...)

The override was missing the 'quantity' parameter, so the positional
argument from callers was captured by 'reserved_quantity' and then
passed as reserved_quantity= keyword to super().

Commit:
[FIX] stock_move_backdating: fix _update_available_quantity signature

The override renamed the 'quantity' parameter to 'reserved_quantity',
causing all stock receipts to set reserved_quantity instead of on-hand
quantity on quants. This resulted in on-hand=0 for every validated
receipt while reserved_quantity was incorrectly set.

The base signature is:
  _update_available_quantity(product_id, location_id, quantity=False,
    reserved_quantity=False, ...)

The override was missing the 'quantity' parameter, so the positional
argument from callers was captured by 'reserved_quantity' and then
passed as reserved_quantity= keyword to super().

Commit:
[FIX] stock_move_backdating: fix _get_price_unit return value

In Odoo 18, _get_price_unit returns {lot: price} dict. The override
always wrapped the result in {lot: value}, even when the backdating
conditions were not met - causing a nested dict {lot: {lot: price}}
that broke standard_price writes downstream.

Now only wraps in dict when the price is actually recalculated for
backdating. Otherwise returns super()'s result as-is.

Commit:
[FIX] stock_move_backdating: fix _get_price_unit return value

In Odoo 18, _get_price_unit returns {lot: price} dict. The override
always wrapped the result in {lot: value}, even when the backdating
conditions were not met - causing a nested dict {lot: {lot: price}}
that broke standard_price writes downstream.

Now only wraps in dict when the price is actually recalculated for
backdating. Otherwise returns super()'s result as-is.

Commit:
[FIX] stock_move_backdating: use LazyTranslate in check_date

check_date is a module-level function without access to self.env.
Using _() caused 'no translation language detected' warnings that
the CI error scanner treats as failures.

Use LazyTranslate instead, which defers translation lookup to render
time when the language context is available.

bosd commented:
:green_circle: Ready now

rousseldenis commented:
/ocabot migration stock_move_backdating

Saran440 reviewed (approved):
Functional Test 👍 

Note: Can you rebase your fixes into `[MIG] stock_move_backdating`, please?

rousseldenis commented on stock_move_backdating/models/stock_move_line.py:
```diff
@@ -2,14 +2,19 @@
     if date and date > now:
-        raise UserError(_("You can not process an actual movement date in the future."))
+        raise UserError(
+            _lt("You can not process an actual movement date in the future.")
```
@bosd You should not use lazy translations in code like this. Odoo specified it should only used with globally defined constants: https://github.com/odoo/odoo/commit/b794f0f332f473deb2c04eba60baf4761db3b508

rousseldenis replied:
Please use `self.env._()` instead

bosd commented:
@rousseldenis Can you please merge?
We would like to start on the translation.

rousseldenis commented:
/ocabot merge nobump
